### PR TITLE
51 implement download functionality

### LINKED ISSUE #51: Implement the download functionality
The student click on the download button when he is watching a lecture and the video gets downloaded as a mp4 file. 
- [x] A reference to the downloaded video should be saved so the video can be deleted or accessed again
- [x] #189

### LINKED ISSUE #186: Implement a new or make changes in old video player to play downloaded videos

### LINKED ISSUE #201: Delete all pop up when a user logs out

### LINKED ISSUE #203: Configure downloads UI
When there are no downloaded courses, it still has an old design. Remove refresh button, add title ,add No downloaded courses

### LINKED ISSUE #224: [Bug] refresh downloads, remove all downloads from UI

### PR DESCRIPTION
**Pull request by Anishyou**
0. Implement Download Functionality
1. Offline video player: https://github.com/TUM-Dev/gocast_mobile/issues/186
2. Implement Delete functionality: https://github.com/TUM-Dev/gocast_mobile/issues/189
3. https://github.com/TUM-Dev/gocast_mobile/issues/203
4. https://github.com/TUM-Dev/gocast_mobile/issues/201

### CODE REVIEW

File: lib/views/video_view/offline_video_player/offline_video_player.dart [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+  bool _shouldMarkAsWatched(double progress) {
+    const watchedThreshold = 0.9; // 80%
+    return progress >= watchedThreshold;
+  }
+
+  Future<void> _markStreamAsWatched() async {
+    final prefs = await SharedPreferences.getInstance();
+    await prefs.setBool('watched_${widget.localPath}', true);
+  }
```
Comment (carlobortolan) [inline lib/views/video_view/offline_video_player/offline_video_player.dart:151]: What do we need this code for? (isn't mark as watched only needed for live streams and VoDs?)

**Review by carlobortolan — CHANGES_REQUESTED** [review]
Good job with the downloads. :+1: 
Here's what I noted:
- There are still quite a lot of **placeholders** which make the downloads confusing to use (see video) - especially because you don't know which lectures you downloaded as they don't have the lecture's or course's names

- When downloading a video on android it **doesn't show any progress/loading** or alert that shows that a video has been successfully downloaded

- The **downloaded videos disappear when refreshing** the "Downloads" page (see video)

https://github.com/TUM-Dev/gocast_mobile/assets/106114526/57645c7a-ca56-49a1-b3c7-2067031f996d

File: lib/view_models/stream_view_model.dart [hunk 2]
```diff
@@ -7,13 +7,121 @@ import 'package:gocast_mobile/base/networking/api/handler/stream_handler.dart';
...
+        downloadedVideos: Map.from(state.downloadedVideos)
+          ..[state.streams?.first.id.toInt() ?? -1] = filePath,
+      );
+
+      return filePath;
+    } catch (e) {
+      _logger.e("Error downloading video: $e");
+      return '';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Anishyou) [reply lib/view_models/stream_view_model.dart:39]: Yes, I have not added that. Thanks for reminding me. I can create another task for it. :)

File: lib/views/video_view/offline_video_player/offline_video_player.dart [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    return position.inSeconds / duration.inSeconds;
+  }
+
+  Future<void> _updateProgress(double progress) async {
+    final prefs = await SharedPreferences.getInstance();
+    await prefs.setDouble('progress_${widget.localPath}', progress);
+  }
+
+  bool _shouldMarkAsWatched(double progress) {
+    const watchedThreshold = 0.9; // 80%
```
[the remark this answers is not in the document]
Comment (Anishyou) [reply lib/views/video_view/offline_video_player/offline_video_player.dart:144]: i think it is 90% I will update the comment

File: lib/view_models/stream_view_model.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Anishyou) [reply lib/view_models/stream_view_model.dart]: Hi, I pushed the recent code with some notification. So it is more clear. We cana dd some better notification loading etc . I will do it in next PR

**Comment by Anishyou** [comment]
When downloading a video on android it doesn't show any progress/loading or alert that shows that a video has been successfully downloaded. 👍 

The downloaded videos disappear when refreshing the "Downloads" page (see video) 👍 

I am planning to do the placeholder in next PR.

**Review by carlobortolan — CHANGES_REQUESTED** [review]
The reloading issue still isn't solved yet, as when you download a video, go to downloads, go back to the video and back to downloads,  the video disappears. 
 
https://github.com/TUM-Dev/gocast_mobile/assets/106114526/b270f8be-026a-409b-a224-18b718d33d15

**Comment by Anishyou** [comment]
I solved the main issue. It was my fault. I was using streamviewmodel to download. It created stateproblem as when you click on a video it refreshes the whole thing. Thanks Carlo, Great work!

**Review by carlobortolan — APPROVED** [review]

**Pull request merged**